server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java

```diff
@@ -140,13 +143,17 @@ public void execute(SearchContext context) {
         }
 
         try {
+            int[] docIds = Arrays.copyOfRange(context.docIdsToLoad(), context.docIdsToLoadFrom(), context.docIdsToLoadSize());
+            int[] sortedDocIds = docIds.clone();
+            Arrays.sort(sortedDocIds);
+
```

Comment:
Sorry I wasn't clear. I was more thinking of something like this:
`````
static class DocIdAndIndex implements Comparable<DocIdAndIndex> {
        final int docId;
        final int index;

        DocIdAndIndex(int docId, int index) {
            this.docId = docId;
            this.index = index;
        }


        @Override
        public int compareTo(DocIdAndIndex o) {
            return Integer.compare(docId, o.docId);
        }
}
....
  DocIdAndIndex[] docs = new DocIdAndIndex[context.docIdsToLoadSize()]; 
  for (int index = 0; index < context.docIdsToLoadSize(); index++) {
    docs[index] = new DocIdAndIndex(context.docIdsToLoad()[context.docIdsToLoadFrom() + index], index);
  }
  Arrays.sort(docs)
`````

You can then use `docs` to retrieve the original index and you don't have the array twice ?

Reply:
Thanks @jimczi . I have one question here. If we use a custom comparator, the performance of timSort or quicksort is O(nlogn).but the complexity of constructing a hashmap is O(n). Maybe we can have a better performance if we use the latter?

Reply:
I don't see how you'd avoid the initial sort by doc ids ? The proposed solution requires to sort the array once and avoids building an hashmap, isn't it better ?

Reply:
In [the first commit](https://github.com/elastic/elasticsearch/pull/57273/commits/65132dfd9bf1f71a17cc2be2033836bb0b9f3bfc), I built the hashmap with the array without sorting it. It only took O(n) to iterate the array once.

Reply:
We still need to provide the array of `SearchHit` sorted by doc ids to the sub fetch phase (hitsExecute) so I don't see how the hashmap would be enough. The current change allows to sort the array once before executing the sub fetch phase so it's an enhancement. Also note that the array is limited to 10k by default since we have a soft limit for the number of hits that can retrieved.